govcd/vm.go

```diff
@@ -164,53 +205,80 @@ func (vm *VM) ChangeCPUCountWithCore(virtualCpuCount int, coresPerSocket *int) (
-					// TODO: Check a valid IP has been given
-					ipAddress = ip
-				} else if network["ip"].(string) != "" {
+
+				// Legacy ip="valid_ip" mode (currently it is hit by ip_allocation_mode=MANUAL as well)
+				case ipIsSet && net.ParseIP(ipFieldString) != nil:
 					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = network["ip"].(string)
-				} else if ip == "" {
+					ipAddress = ipFieldString
+				case ipIsSet && ipField != "": // Legacy ip="something_invalid" we default to DHCP. This is odd but backwards compatible.
 					ipAllocationMode = types.IPAllocationModeDHCP
+					// TODO 3.0 remove until here when legacy `ip` and `network_name` attributes are removed
+
+				case ipIsSet && net.ParseIP(ipFieldString) != nil && (network["ip_allocation_mode"].(string) == types.IPAllocationModeManual):
+					ipAllocationMode = types.IPAllocationModeManual
+					ipAddress = ipFieldString
+				default: // New functionality. IP was not set and we're defaulting to provided ip_allocation_mode (only manual requires the IP)
+					ipAllocationMode = network["ip_allocation_mode"].(string)
 				}
 
-				util.Logger.Printf("[DEBUG] Function ChangeNetworkConfig() for %s invoked", network["orgnetwork"])
+				networkSection.NetworkConnection[loopIndex].NeedsCustomization = true
+				networkSection.NetworkConnection[loopIndex].IsConnected = true
+				networkSection.NetworkConnection[loopIndex].IPAddress = ipAddress
+				networkSection.NetworkConnection[loopIndex].IPAddressAllocationMode = ipAllocationMode
 
-				networkSection.NetworkConnection[index].NeedsCustomization = true
-				networkSection.NetworkConnection[index].IPAddress = ipAddress
-				networkSection.NetworkConnection[index].IPAddressAllocationMode = ipAllocationMode
+				// for IPAllocationModeNone we hardcode special network name used by vcd 'none'
+				if ipAllocationMode == types.IPAllocationModeNone {
+					networkSection.NetworkConnection[loopIndex].Network = types.NoneNetwork
+				} else {
+					networkSection.NetworkConnection[loopIndex].Network = network["network_name"].(string)
+				}
 
-				if network["is_primary"] == true {
-					networkSection.PrimaryNetworkConnectionIndex = index
+				// If we have one NIC only then it is primary by default, otherwise we check for "is_primary" key
+				if (len(networks) == 1) || (network["is_primary"] != nil && network["is_primary"].(bool)) {
+					networkSection.PrimaryNetworkConnectionIndex = tfNicSlot
 				}
 			}
 		}
 	}
+}
+
+func (vm *VM) ChangeNetworkConfig(networks []map[string]interface{}) (Task, error) {
+	err := vm.Refresh()
+	if err != nil {
+		return Task{}, fmt.Errorf("error refreshing VM before running customization: %v", err)
+	}
+
+	// The API returns unordered list of NICs. This means that networkSection.NetworkConnection[0] will not
+	// necessarily be NIC 0 if there is more than one NIC available.
+	networkSection, err := vm.GetNetworkConnectionSection()
```

Comment:
This is a useful comment, but why do we have it here is not very clear. After reading such comment I am instinctively hoping to see some reordering magic lines :) If they are somewhere here, then it's probably worth elaborating in this comment as to how it is done.

Reply:
Absolutely right. Over the time this sorting got to other place.